packages/tpetra/core/src/Tpetra_FECrsMatrix_def.hpp

```diff
@@ -138,8 +140,159 @@ void FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::beginFill()  {
   this->resumeFill();
 }
 
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::beginAssembly() {
+  const char tfecfFuncName[] = "FECrsMatrix::beginAssembly: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::closed,
+    std::runtime_error,
+    "Cannot beginAssembly, matrix is not in a closed state"
+  );
+  *fillState_ = FillState::open;
+  this->beginFill();
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::endAssembly() {
+  const char tfecfFuncName[] = "FECrsMatrix::endAssembly: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::open,
+    std::runtime_error,
+    "Cannot endAssembly, matrix is not open to fill."
+  );
+  *fillState_ = FillState::closed;
+  this->endFill();
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::beginModify() {
+  const char tfecfFuncName[] = "FECrsMatrix::beginModify: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::closed,
+    std::runtime_error,
+    "Cannot beginModify, matrix is not in a closed state"
+  );
+  *fillState_ = FillState::modify;
+  this->resumeFill();
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::endModify() {
+  const char tfecfFuncName[] = "FECrsMatrix::endModify: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::modify,
+    std::runtime_error,
+    "Cannot endModify, matrix is not open to modify."
+  );
+  *fillState_ = FillState::closed;
+  this->fillComplete();
+}
+
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+LocalOrdinal
+FECrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::replaceGlobalValuesImpl(
+  impl_scalar_type rowVals[],
+  const crs_graph_type& graph,
+  const RowInfo& rowInfo,
+  const GlobalOrdinal inds[],
+  const impl_scalar_type newVals[],
+  const LocalOrdinal numElts) const
+{
+  const char tfecfFuncName[] = "FECrsMatrix::replaceGlobalValues: ";
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
+    *fillState_ != FillState::open,
+    std::runtime_error,
+    "Cannot replace global values, matrix is not open to fill."
+  );
```

Comment:
Why does this have a different behavior than replaceLocalValues?

Reply:
replaceGlobalValues does not allow replacing global values during the modify phase because globalAssemble is not called after endModify like it is after endAssembly.

Reply:
Is that consistent with regular CrsMatrix (and how it handles off-processor replacement)?

Reply:
@csiefer - no it's not really the same.  With vanilla CrsMatrix you have to do all the import/export wizardry yourself.  FECrsMatrix does it for  you - with caveats.